### PR TITLE
Check to ensure sample IDs match MLST.JSON keys

### PR DESCRIPTION
**Pull request by kylacochrane**
This PR verifies that the sample IDs in the input samplesheet align correctly with their corresponding keys in the MLST JSON file.

- `INPUT_CHECK` - verifies the sample ID and JSON key matches; outputs and maps the boolean value to the sample metadata; generates error_report.csv if applicable

NOTE: See https://github.com/phac-nml/gasnomenclature/pull/10 for the addition of integration tests

### CODE REVIEW

**Review by apetkau — CHANGES_REQUESTED** [review]
This is amazing. Thanks so much @kylacochrane 😄 .

In addition to what we talked about I had a few other comments.

File: workflows/gas_nomenclature.nf [hunk 1]
```diff
@@ -68,13 +71,33 @@ workflow GAS_NOMENCLATURE {
...
+    // Ensure meta.id and mlst_file keys match; generate error report for samples where id ≠ key
+    id_key = INPUT_CHECK(input)
+    ch_versions = ch_versions.mix(id_key.versions)
+
+    error_report = ERROR_REPORT(input)
+    ch_versions = ch_versions.mix(error_report.versions)
+
+    // Update metadata to include the id_key.match data
+    match = id_key.match.map { meta, file, json ->
+        def id_match = file.text.trim()
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply workflows/gas_nomenclature.nf:85]: I am actually quite surprised it did work. Although I have made a note that we may want to switch to a Groovy code in the future to update the `meta` object instead 🤔

File: modules/local/filter_query/main.nf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def out_extension = out_format == "tsv" ? 'tsv' : 'csv'
+
+    """
+    # Filter the query samples only; keep only the 'id' and 'address' columns
+    csvtk filter2 \\
+        ${addresses} \\
+        --filter '\$id == \"$queryID\"' \\
+        --delimiter "${delimiter}" \\
+        --out-delimiter "${out_delimiter}" \\
+        --out-file ${outputFile}.tmp
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply modules/local/filter_query/main.nf]: Oh yes! Sure thing: [6cf40c3](https://github.com/phac-nml/gasnomenclature/pull/9/commits/6cf40c30d6aed50d13cb8c151cd94ec5236f8ecc)

File: modules/local/filter_query/main.nf [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    csvtk cut -f 1,2 ${outputFile}.tmp > ${outputFile}.${out_extension}
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply modules/local/filter_query/main.nf]: indeed 😄 [6cf40c3](https://github.com/phac-nml/gasnomenclature/pull/9/commits/6cf40c30d6aed50d13cb8c151cd94ec5236f8ecc)

File: bin/check_samplesheet.py [hunk 4]
```diff
@@ -195,7 +195,9 @@ def check_samplesheet(file_in, file_out):
         # Validate the existence of the expected header columns.
         if not required_columns.issubset(reader.fieldnames):
             req_cols = ", ".join(required_columns)
-            logger.critical(f"The sample sheet **must** contain these column headers: {req_cols}.")
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply bin/check_samplesheet.py:198]: Removed here: [5bfe569](https://github.com/phac-nml/gasnomenclature/pull/9/commits/5bfe569ec975ee4b38dc91adec005541f9f314a4)

File: bin/error_report.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+
+import json
+import argparse
+import csv
+
+
+def check_inputs(json_file, sample_id, address, output_file):
+    # Define a variable to store the match_status (True or False)
+    json_data = json.load(open(json_file))
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply bin/error_report.py]: Done here: [cf37bfe](https://github.com/phac-nml/gasnomenclature/pull/9/commits/cf37bfea9b84bb97ae5df045aee57ff98a3f8d91)

**Comment by kylacochrane** [comment]
**Additional changes:**
Combined INPUT_CHECK and ERROR_REPORT into a single process: [cf37bfe](https://github.com/phac-nml/gasnomenclature/pull/9/commits/cf37bfea9b84bb97ae5df045aee57ff98a3f8d91)
Filtered out samples using Groovy logic in workflow (instead of a separate process): [5456aa7](https://github.com/phac-nml/gasnomenclature/pull/9/commits/5456aa7050dbb7d71387f7a70d2da302b9f5635f)

File: workflows/gas_nomenclature.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (apetkau) [reply workflows/gas_nomenclature.nf:80]: Thanks so much for testing this out. I'm glad it does work 😄

**Review by apetkau — CHANGES_REQUESTED** [review]
This looks great. Thanks so much for your work @kylacochrane 😄 . Just one small comment.

File: modules/nf-core/custom/dumpsoftwareversions/templates/dumpsoftwareversions.py [hunk 6]
```diff
@@ -58,7 +58,9 @@ def main():
     }
 
     with open("$versions") as f:
-        versions_by_process = yaml.load(f, Loader=yaml.BaseLoader) | versions_this_module
+        versions_by_process = (
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply modules/nf-core/custom/dumpsoftwareversions/templates/dumpsoftwareversions.py]: Reverted in: [4589456](https://github.com/phac-nml/gasnomenclature/pull/9/commits/4589456a2cc7117adf54b95831a80925bd653d11)

File: workflows/gas_nomenclature.nf [hunk 7]
```diff
@@ -68,13 +69,34 @@ workflow GAS_NOMENCLATURE {
...
+    input = Channel.fromSamplesheet("input")
+
+    // Ensure meta.id and mlst_file keys match; generate error report for samples where id ≠ key
+    id_key = INPUT_CHECK(input)
+    ch_versions = ch_versions.mix(id_key.versions)
+
+    // Update metadata to include the id_key.match data
+    match = id_key.match.map { meta, file, json ->
+        def id_match = file.text.trim()
+        [meta + [id_match: id_match], json]
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply workflows/gas_nomenclature.nf]: Done here: [9748287](https://github.com/phac-nml/gasnomenclature/pull/9/commits/97482872ab05664a0235709aa7962ee8e658edb2)

File: tests/data/samplesheets/samplesheet_test1.csv [hunk 8]
```diff
@@ -0,0 +1,5 @@
+sample,mlst_alleles,address
+sampleQ,https://raw.githubusercontent.com/phac-nml/gasnomenclature/add_tests/tests/data/reports/sampleF.mlst.json,
+sample1,https://raw.githubusercontent.com/phac-nml/gasnomenclature/dev/tests/data/reports/sample1.mlst.json,1.1.1
+sample2,https://raw.githubusercontent.com/phac-nml/gasnomenclature/add_tests/tests/data/reports/sample7.mlst.json,1.1.1
```
Comment (apetkau) [inline tests/data/samplesheets/samplesheet_test1.csv:4]: As a note, once this PR is merged you'll have to update the URL here to point to `dev`. You can just directly commit that to the `dev` branch.

**Review by apetkau — APPROVED** [review]
This looks great. Thanks so much for all your hard work on this @kylacochrane 😄

**Pull request merged**